x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -805,65 +835,28 @@ private void innerRefresh(String tokenDocId, Map<String, Object> source, long se
-                            parsedTokens = parseTokensFromDocument(response.getSource(), null);
-                        } catch (IllegalStateException | DateTimeException e) {
-                            logger.error("unable to decode existing user token", e);
-                            listener.onFailure(new ElasticsearchSecurityException("could not refresh the requested token", e));
-                            return;
-                        }
-                        listener.onResponse(parsedTokens);
-                    } else {
-                        // We retry this since the creation of the superseding token document might already be in flight but not
-                        // yet completed, triggered by a refresh request that came a few milliseconds ago
-                        logger.info("could not find superseding token document [{}] for token document [{}], retrying",
-                                refreshTokenStatus.getSupersedingDocId(), tokenDocId);
-                        maybeRetryOnFailure.accept(invalidGrantException("could not refresh the requested token"));
-                    }
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    if (isShardNotAvailableException(e)) {
-                        logger.info("could not find superseding token document [{}] for refresh, retrying",
-                                refreshTokenStatus.getSupersedingDocId());
-                        maybeRetryOnFailure.accept(invalidGrantException("could not refresh the requested token"));
-                    } else {
-                        logger.warn("could not find superseding token document [{}] for refresh", refreshTokenStatus.getSupersedingDocId());
-                        onFailure.accept(invalidGrantException("could not refresh the requested token"));
-                    }
-                }
-            });
+                    tokenDocId, refreshTokenStatus.getSupersededBy());
+            getSupersedingTokenDocAsyncWithRetry(refreshTokenStatus, backoff, listener);
         } else {
             final String newUserTokenId = UUIDs.randomBase64UUID();
+            final Version newTokenVersion = getTokenVersionCompatibility();
             final Map<String, Object> updateMap = new HashMap<>();
             updateMap.put("refreshed", true);
             updateMap.put("refresh_time", clock.instant().toEpochMilli());
-            updateMap.put("superseded_by", getTokenDocumentId(newUserTokenId));
+            if (newTokenVersion.onOrAfter(VERSION_TOKENS_INDEX_INTRODUCED)) {
+                // the superseding token document reference is formated as: "<alias>|<document_id>" ; the alias points to a single index
+                // containing the document with the said id
+                updateMap.put("superseded_by",
+                        getTokensIndexForVersion(newTokenVersion).aliasName() + "|" + getTokenDocumentId(newUserTokenId));
```

Comment:
I feel like this code, and the code to parse the superseded_by should be consistent about whether to use a fixed alias name or not.
This code implies that it's possible to use this `alias|id` format with either index, but the parsing code only supports reading from the token index.
I don't mind which way we go, but it's easier to reason about if they make the same assumptions.

Reply:
I made it clear in the comments, as well as in the code, that the only valid format of the `superseded_by` reference is `.security-tokens|<doc_id>` . This code was faking to be generic with regards to the tokens alias.